### PR TITLE
WIP - Create Developers Map

### PR DESCRIPTION
**Pull request by doodlesbykumbi**

### CODE REVIEW

File: developers-map.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
++ represents the secretless configuration as struct
+
+### Entrypoint
+[./cmd/secretless-broker/main.go](./cmd/secretless-broker/main.go)
+obtains the following via flags:
+
++ -f secretless config file
++ -p plugin directory
++ --watch condition to reload when config file changes
++ --debug to provide debug information
```
Comment (izgeri) [inline developers-map.md]: Missing `config-mgr` here

File: developers-map.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
+# Developer Map
+
+This document is meant to provide a mapping from the conceptual model of Secretless to the codebase.
+
+## Directories
+
++ [./cmd](./cmd) - entrypoints for [binaries](#binaries)
++ [./internal/app](./internal/app) - internal implementations of [plugins](#plugins)
++ [./internal/pkg](./internal/pkg) - [plugin manager](#pluginmanager) and [resolver](#resolver)
++ [./pkg/secretless](./pkg/secretless) - secretless [config](#config) and [plugin interfaces](#plugins)
```
Comment (sgnn7) [inline developers-map.md]: pkg/apis/secretless.io for k8s CRDs codegen

File: developers-map.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+### Configuration
+config.Config
+[./pkg/secretless/config/config.go](./pkg/secretless/config/config.go)
+
++ represents the secretless configuration as struct
+
+### Entrypoint
+[./cmd/secretless-broker/main.go](./cmd/secretless-broker/main.go)
+
+At the entrpoiny, some parameters are obtained via flags:
```
Comment (sgnn7) [inline developers-map.md]: entrypoint

File: developers-map.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
++ -f secretless config file
++ -p plugin directory
++ --watch condition to reload when config file changes
++ --debug toggles debug information
++ --config-mgr strategy to provide secretless configuration
```
Comment (sgnn7) [inline developers-map.md]: iff you don't want to use the default config-file configuration manager

File: developers-map.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+Execution shifts to the [Plugin Manager](#pluginmanager):
+  + load internal [plugins](#plugins)
+  + load external [plugins](#plugins) from the plugin directory
+  + registers signal handlers
+  + run
```
Comment (sgnn7) [inline developers-map.md]: Would be neat if these were links to sections of code that does it

File: developers-map.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
++ root of
+  + RunHandlerFunc
+  + RunListenerFunc
++ manages all the [plugins](#plugins) 
++ initialises and runs [Proxy](#proxy) 
++ implements [EventNotifier](#eventnotifier) (only one in the whole repo)
+
+
+## Plugins
+Plugins are used as a mechanism for extending Secretless functionality beyond the core. These are expressed as interfaces. They are versioned in anticipation of growth.
```
Comment (sgnn7) [inline developers-map.md]: To be more exact, plugins are used also to provide core functionality - without internal plugins, secretless by itself doesn't work.

File: developers-map.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
++ v1 interfaces are defined at [./pkg/secretless/plugin/v1](./pkg/secretless/plugin/v1)
++ list of interfaces
+  + ConfigurationManager 
+    - pushes configuration data and triggers updates
+    - interface: [./pkg/secretless/plugin/v1/configuration_manager.go](./pkg/secretless/plugin/v1/configuration_manager.go)
+    - implementation: [./internal/app/secretless/configurationmanagers/configfile](./internal/app/secretless/configurationmanagers/configfile)
```
Comment (sgnn7) [inline developers-map.md]: s/implementation/implementation example/

Comment (sgnn7) [reply developers-map.md]: ditto on the rest in the list

File: developers-map.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  + ConnectionManager
+    - manages connections for handlers and listeners
+    - interface: ./pkg/secretless/plugin/v1/connection_manager.go
```
Comment (sgnn7) [inline developers-map.md]: needs to be a link

File: developers-map.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    - interface: [./pkg/secretless/plugin/v1/listener.go](./pkg/secretless/plugin/v1/listener.go)
+    - pg implementation: [./internal/app/secretless/listeners/pg/listener.go](./internal/app/secretless/listeners/pg/listener.go)
+  + [Handler](#handler)
+    - receives client connection, connect it to a backend and streams connection.
+    - interface: [./pkg/secretless/plugin/v1/handler.go](./pkg/secretless/plugin/v1/handler.go)
+    - pg implementation: [./internal/app/secretless/handlers/pg/handler.go](./internal/app/secretless/handlers/pg/handler.go)
+  + Provider
+    - used to obtain values from a secret vault backend
+    - interface: [./pkg/secretless/plugin/v1/provider.go](./pkg/secretless/plugin/v1/provider.go)
+    - env implementation: [./internal/app/secretless/providers/env/provider.go](./internal/app/secretless/providers/env/provider.go)
```
Comment (sgnn7) [inline developers-map.md]: s/env/\'env\'/

File: developers-map.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+  + [Resolver](#resolver)
+    - manages Providers and provides convenient interface to obtain multiple values from multiple secret vault backends
+    - interface: [./pkg/secretless/plugin/v1/resolver.go](./pkg/secretless/plugin/v1/resolver.go)
+    - implementation: [./internal/pkg/plugin/resolver.go](./internal/pkg/plugin/resolver.go)
+ 
++ internal implementations of most plugin interfaces
+  + located at ./internal/app/secretless
```
Comment (sgnn7) [inline developers-map.md]: Needs to be a link

File: developers-map.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+  + most plugins are exposed as factories, where the key is the identifier and the value is an interface pointer e.g. map[string]func(plugin_v1.HandlerOptions) plugin_v1.Handler
```
Comment (sgnn7) [inline developers-map.md]: Escape the code w/ backticks

File: developers-map.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
++ --debug toggles debug information
++ --config-mgr strategy to provide secretless configuration
+
+Execution shifts to the [Plugin Manager](#pluginmanager):
+  + load internal [plugins](#plugins)
+  + load external [plugins](#plugins) from the plugin directory
+  + registers signal handlers
+  + run
+
+### PluginManager
```
Comment (sgnn7) [inline developers-map.md]: Most itemized items below this point in the file need to be capitalized

File: developers-map.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+  + RunHandlerFunc - HandlerFactory, passed to listener
+    + NOTE: HandlerFactories run Handlers before returning them.
+    + takes 
+      + handler id
+      + HandlerOptions
+    + returns Listener interface pointer
++ manages Listener lifecycles
+  + in #Listen, for each listener 
+    + creates net.Listener based on config.Config
+    + creates Listener interface pointer using RunListenerFunc
```
Comment (sgnn7) [inline developers-map.md]: consider for later doing this with some generator or the docs will get out of date quick

File: developers-map.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+### Resolver
+interface: [./pkg/secretless/plugin/v1/resolver.go](./pkg/secretless/plugin/v1/resolver.go)
+
+implementation: [./internal/pkg/plugin/resolver.go](./internal/pkg/plugin/resolver.go)
+
++ manages Providers and provides convenient interface to obtain multiple values from multiple secret vault backends
++ takes
+  + ProviderFactories
+  + EventNotifier
+
```
Comment (sgnn7) [inline developers-map.md]: extra newline not needed

**Pull request closed without merging**